Propose voting over waku protocol (#1)
Szymx95/rfc: content/docs/rfcs/24/README.md

---
slug: 24

> title: 24/WAKU-VOTING

Reviewer: ```suggestion
title: 24/STATUS-CURATION
```

For now, let's keep the spec focus on our use case: Status Community Directory Curation DApp.
Someone else can specify a more generic protocol when they need it.

> name: Waku v2 Voting

Reviewer: ```suggestion
name: Status Community Directory Curation Voting using Waku v2
```

> status: raw

Reviewer: ```suggestion
status: raw
category: application
```
We have started to add this metadata field. See https://github.com/vacp2p/rfc/issues/368.

editor: Szymon Szlachtowicz <szymon.s@ethworks.io>
---

This spec is a proposition for a voting protocol over Waku V2.

# Motivation

In open p2p protocol there is an issue with voting off-chain as there is much room for malicious peers to only include votes that support their case when submitting votes to chain.

Proposed solution is to agregate votes over waku and allow users to submit votes to smart contract that aren't already submitted.

# Smart contract

> Voting should be finalized on chain so that the finished vote is immutable. Becouse of that, smart contract needs to be deployed. New smart contract will have to be able to start a vote and accept submited votes.

Reviewer: ```suggestion
Voting should be finalized on chain so that the finished vote is immutable.
Because of that, smart contract needs to be deployed.
New smart contract will have to be able to start a vote and accept submitted votes.
```
Please use semantic break lines https://sembr.org/

> Voting should be finalized on chain so that the finished vote is immutable. Becouse of that, smart contract needs to be deployed. New smart contract will have to be able to start a vote and accept submited votes.

Reviewer: So the idea is to deploy one smart contract per vote?

Author: No, a smart contract for the application i will try to write it more clearly.

## Double voting

Smart contract should also keep a list of all votes so that no one can double vote.

# Voting

## Sending votes

Sending votes is simple every peer is able to send a message to Waku topic specific to given application: 
```

> /{application-name}/{voting-room}

Reviewer: ```suggestion
/status-community-directory-curation-vote/1/{room-identifier}/proto
```
See https://rfc.vac.dev/spec/23/: `/{application-name}/{version-of-the-application}/{content-topic-name}/{encoding}`.

In the spirit on having specific to our use case (Status Community Directory Curation DApp) then I believe the proposed content topic makes more sense.

Should `room-identifier` be the room's public key?

Also, I assume the usage of protobuf, hence `proto`. Is that fine?

Author: After some thought i think that we can simply use incremental id of voting room [see here](https://github.com/Szymx95/rfc/pull/1#discussion_r647170066) instead of room identifier. Since all the voting rooms will be public we don't have to worry about someone guessing the voting room id

```

vote object that is sent over waku should contain information about: 

```ts
{

>     sender: String //address of a sender

Reviewer: ```suggestion
    sender: String // address of the sender
```

>     vote: String //vote sent eg. 'yes' 'no'

Reviewer: ```suggestion
    vote: String // vote sent eg. 'yes' 'no'
```

>     sntAmount: BigNumber //number of snt cast on vote

Reviewer: ```suggestion
    sntAmount: BigNumber // number of SNT cast on vote
```

>     sign: String // cryptographic sign of a transaction

Reviewer: ```suggestion
    sign: String // cryptographic signature of a transaction
```

What will be signed? Which fields?

>     nonce: Number // number of votes cast from this address on current vote

Reviewer: What do you mean? Are you saying that a single voter may decide to vote several time? (ie, to increase the number of token they cast on the vote).

Author: That was my idea, that maybe someone may want to add more SNT later or have a change of heart and vote differently

Reviewer: I think it's a nice to have, I would not build this feature in the first version.

>     room: String // address of a voting room

Reviewer: Address or public key?

Also, if a room is voted in (vote A), then removed (vote B), then a new vote is started (vote C), how do we ensure that nobody is going to re-use the messages sent during A?

I believe we need some vote epoch identifier, so that votes are not re-used.

Author: My idea was that room is a public key of voting room which is a specific vote and smart contract would hold information of which community is voted on and what type of vote is being. made (addition, deletion).

But i think that simply having a id of vote is enough.

So for example we want to add new community to directory `Awesome Community`. So we create a new voting room by sending a public key (0x231...) of community and vote type ('add') to smart contract and in turn we get a id of this voting room (e.g. 23). Then when we want to vote on this community we just pass in the id of voting room.

Then when we want to remove this community from directory we start a new voting room which will get different id and votes can't be reused.

}
```

## Aggregating votes

> Every peer that is opening specific voting room will listen to votes sent over p2p network, and aggregate them for a single transaction to chain

Reviewer: ```suggestion
Every peer that is opening specific voting room will listen to votes sent over p2p network, and aggregate them for a single transaction to chain.
```

What we discussed was:
- Alice open vote for room 1, she will vote yes so she listens to aggregate yes votes and publshes her yes vote on Waku.
- Bob is made aware of vote for room 1, he votes no, he publishes his no vote on Waku, retrieve previous no votes and starts listening to no votes.
- Carole is made aware of vote for room 1, she votes yes, she publishes her yes vote on Waku, retrieve previous yes votes and starts listening to yes votes.
- Bob really does not want this community to be in the directory, he aggregates no votes and submit to the blockchain
- Carole is really excited about this community, she aggregates yes votes and submit to the blockchain

Author: Yes but as someone pointed out i think it was Marek typical user isn't really aware of what is going under the hood and he doesn't need to know if he is aggregating all the votes or only no or yes votes, he only needs to know if aggregated votes would change the outcome.

 In your example above Bob would see that submitting votes to blockchain would be against his wishes so he doesn't do that but Carole sees that if she submits the votes she can change the results in her favour so she submits votes for and against because she is a good user. That way we can save some gas for users, and create a simpler aggregator.

Now if someone wanted to be malicious and only aggregate votes against normal user would see that only votes for are left and he can simply submit all the votes he has aggregated which are votes for.

Reviewer: > if aggregated votes would change the outcome.

So yes, we discussed that.

I think if we are adding conditionals such as "only show to the user the option to submit if it can change the outcome" we could end up degrading the UX.

Maybe a simplest way to proceed would be to simply show to the user:
"4000 SNT voted yes, 3000 voted no. There are 1000 pending yes vote. Do you want to submit them?".

This way, we don't have to fiddle with conditionals based on number of votes and last minute voting, etc.

Author: So maybe we could just show current submitted votes and under this show how the votes would look like after submission for example:

"4000 SNT voted yes, 3000 SNT voted no. There are pending votes. Do you want to submit them?".

And if someone would click submit button modal would show up with information how votes would look like after submission:

"Votes after submission 5000 SNT yes, 6000 SNT no."

This way we have simple logic of aggregating all votes with no need to track what user voted before and filtering votes

## Submiting to chain

> Every peer that has aggregated at least one vote will be able to send them to smart contract.

Reviewer: The peer's vote count too, a vote does not need to be aggregated to be submitted.

Author: I know that. But peer also aggregates his own votes, so when he casts a vote he has aggregated one vote and can submit that

Smart contract needs to verify that all votes are valid (eg. all adress had enough SNT, all votes are correctly signed) and that votes aren't duplicated on smart contract.

## Finalizing 

> Once given time window (block of given number was mined) has passed smart contract will no longer accept adding new votes and it will be possible to call smart contract with address of voting room and get back result of voting.

Reviewer: This is vague and does not really matches the product requirements. I'd just leave it as:
 ```suggestion
Once the vote deadline has expired, the smart contract will not accept votes anymore.
```
